### PR TITLE
docs(core/config): config options for advanced users only

### LINKED ISSUE #18408: Flag config options in code to show "for advanced users only" admonition
### What would you like Renovate to be able to do?

Right now, we have to put an admonition in the docs to tell users that a config option is "for advanced users only".

@rarkins had the idea of automating this, like we do already for experimental config options.

### If you have any ideas on how this should be implemented, please tell us here.

- [x] Create new config option Boolean called `advancedUse` that we can attach to config options in the code
- [x] Write code that can create the `warning` admonition for us.
- [x] _After_ the PR lands that makes the code and admonition work: remove the old manual "for advanced users" admonitions from the docs

How it should work:

- If `advancedUse=true` then our docs build process adds the admonition to the docs
- If `advancedUse=false` then nothing happens.

Related comment:

- https://github.com/renovatebot/renovate/issues/18404#issuecomment-1283604679

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by MaronHatoum**
## Changes
Add `advancedUse` entry to RenovateOptionBase.

## Context
Closes https://github.com/renovatebot/renovate/issues/18408

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

File: tools/docs/config.ts [hunk 1]
```diff
@@ -233,10 +234,20 @@ export async function generateConfig(dist: string, bot = false): Promise<void> {
...
       }
     });
 
   await updateFile(`${dist}/${configFile}`, configOptionsRaw.join('\n'));
 }
+
+function generateAdvancedUse(): string {
+  const warning =
+    '\n<!-- prettier-ignore -->\n!!! warning\n    For advanced use only! Use at your own risk!';
+  return warning + '\n';
```
Comment (viceice) [inline tools/docs/config.ts:252]:
```suggestion
    '\n<!-- prettier-ignore -->\n!!! warning\n    For advanced use only! Use at your own risk!\n';
  return warning;
```

no extra string operation 😉

File: docs/usage/configuration-options.md [hunk 2]
```diff
@@ -748,10 +748,6 @@ See [Private module support](https://docs.renovatebot.com/getting-started/privat
 
 ## excludeCommitPaths
 
-<!-- prettier-ignore -->
-!!! warning
-    For advanced users only!
```
Comment (viceice) [inline docs/usage/configuration-options.md:753]: i think there are some more to change. why not doing all in one pr?

Comment (MaronHatoum) [reply docs/usage/configuration-options.md:753]: which changes?

Comment (viceice) [reply docs/usage/configuration-options.md:753]: https://github.com/renovatebot/renovate/blob/867c47ff6542fa36789b574f216f31aa13cdc0a8/docs/usage/configuration-options.md#L396

and below

Comment (viceice) [reply docs/usage/configuration-options.md:753]: https://github.com/renovatebot/renovate/blob/867c47ff6542fa36789b574f216f31aa13cdc0a8/docs/usage/configuration-options.md#L864-L869

https://github.com/renovatebot/renovate/blob/867c47ff6542fa36789b574f216f31aa13cdc0a8/docs/usage/configuration-options.md#L935-L940

https://github.com/renovatebot/renovate/blob/867c47ff6542fa36789b574f216f31aa13cdc0a8/docs/usage/configuration-options.md#L1197-L1201

File: lib/config/types.ts [hunk 3]
```diff
@@ -357,6 +357,16 @@ export interface RenovateOptionBase {
...
+
+  advancedUse?: boolean;
+
+  followTagAdvancedFunctionality?: boolean;
+
+  groupAdvancedFunctionality?: boolean;
+
+  advancedConfig?: boolean;
+
+  commitMsg?: boolean;
```
Comment (HonkingGoose) [inline lib/config/types.ts:369]: Do we usually abbreviate the word `message` to `msg`? I would expect it to be written out fully like this: `commitMessage`.

Also I don't really understand what this Boolean is doing, just by looking at the name `commitMsg`. It looks like we want to nudge our users away from editing the commit message themselves?

How about calling it: `directEditOfCommitMessageDeprecation`? Or at least a more descriptive name. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: tools/docs/config.ts [hunk 4]
```diff
@@ -271,11 +271,18 @@ export async function generateConfig(dist: string, bot = false): Promise<void> {
...
+    '    For advanced use only! Use at your own risk!\n'
+  );
 }
 
 function generateFollowTagAdvancedFunctionality(): string {
-  return "\n<!-- prettier-ignore -->\n!!! warning\n    Advanced functionality.\n    Only use this if you're sure you know what you're doing.\n";
+  return (
+    '\n<!-- prettier-ignore -->\n!!! warning\n' +
+    '    Advanced functionality.\n' +
+    "    Only use this if you're sure you know what you're doing.\n"
```
Comment (HonkingGoose) [inline tools/docs/config.ts:284]:
```suggestion
    '    Only use this if you're sure you know what you're doing.\n'
```

Do you really mean to use `"` here? The other lines use `'`.

**Review by HonkingGoose — COMMENTED** [review]
Asking about single-quote vs double-quote in a piece of code.

File: tools/docs/config.ts [hunk 5]
```diff
@@ -233,10 +237,73 @@ export async function generateConfig(dist: string, bot = false): Promise<void> {
...
+  );
+}
+
+function generateGroupAdvancedFunctionality(): string {
+  return (
+    '\n<!-- prettier-ignore --> \n!!! warning \n' +
+    '    Advanced functionality only. \n' +
+    "   Do not use unless you know what you're doing.\n"
```
Comment (HonkingGoose) [inline tools/docs/config.ts:292]: We need _four_ spaces here for the admonition to work properly.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: tools/docs/config.ts [hunk 6]
```diff
@@ -233,10 +237,65 @@ export async function generateConfig(dist: string, bot = false): Promise<void> {
...
+    '\n<!-- prettier-ignore -->\n!!! warning\n' +
+    '    Advanced config, use at your own risk.\n'
+  );
+}
+
+function generateCommitMsg(): string {
+  return (
+    '\n<!-- prettier-ignore -->\n!!! warning\n' +
+    '    Editing of `commitMessage` directly is now deprecated and not recommended.\n' +
+    '    Please instead edit the fields such as `commitMessageAction`, `commitMessageExtra`, etc.\n'
```
Comment (HonkingGoose) [inline tools/docs/config.ts:299]:
```suggestion
    '    We deprecated editing the `commitMessage` directly, and we recommend you stop using this config option.\n' +
    '    Instead use config options like `commitMessageAction`, `commitMessageExtra`, and so on, to create the commit message you want.\n'
```

How about this wording? We should explain that _we_ deprecated this feature, and what _we_ recommend. Plus we can rewrite the sentences a bit for clarity as well. 😄

File: tools/docs/config.ts [hunk 7]
```diff
@@ -233,10 +237,65 @@ export async function generateConfig(dist: string, bot = false): Promise<void> {
...
+
+function generateAdvancedUse(): string {
+  return (
+    '\n<!-- prettier-ignore -->\n!!! warning\n' +
+    '    For advanced use only! Use at your own risk!\n'
+  );
+}
+
+function generateFollowTagAdvancedFunctionality(): string {
+  return "\n<!-- prettier-ignore -->\n!!! warning\n    Advanced functionality.\n    Only use this if you're sure you know what you're doing.\n";
```
Comment (HonkingGoose) [inline tools/docs/config.ts:281]:
```suggestion
  return "\n<!-- prettier-ignore -->\n!!! warning\n    Advanced functionality.\n    Only use this if you know what you're doing.\n";
```

Is there a reason that `generateFollowTagAdvancedFunctionality` and `generateGroupAdvancedFunctionality` have different string messages? I'm not quite sure what the difference is.

If we need the different messages, then ignore the suggestions that change these strings. 😉

File: tools/docs/config.ts [hunk 8]
```diff
@@ -233,10 +237,65 @@ export async function generateConfig(dist: string, bot = false): Promise<void> {
...
+}
+
+function generateGroupAdvancedFunctionality(): string {
+  return "\n<!-- prettier-ignore --> \n!!! warning\n    Advanced functionality only.\n    Do not use unless you know what you're doing.\n";
```
Comment (HonkingGoose) [inline tools/docs/config.ts:285]:
```suggestion
  return "\n<!-- prettier-ignore --> \n!!! warning\n    Advanced functionality.\n    Only use this if you know what you're doing.\n";
```

File: tools/docs/config.ts [hunk 9]
```diff
@@ -233,10 +237,65 @@ export async function generateConfig(dist: string, bot = false): Promise<void> {
         `\n${option.description}\n\n` +
         genTable(Object.entries(el), option.type, option.default);
 
+      if (el.advancedUse) {
```
Comment (HonkingGoose) [inline tools/docs/config.ts:240]: How do we intend to test if the code works? In production, via code review? I don't see any tests, but maybe it's hard to write test for code like this?

**Review by viceice — CHANGES_REQUESTED** [review]

File: tools/docs/config.ts (hunk 7 above)
Comment (MaronHatoum) [reply tools/docs/config.ts:281]: I think `generateFollowTagAdvancedFunctionality` and `generateGroupAdvancedFunctionality`  should return the same string.

File: tools/docs/config.ts (hunk 9 above)
Comment (MaronHatoum) [reply tools/docs/config.ts:240]: I already ran the `build:docs` command and compared the previous docs with the new ones.
As I see there are no tests for this right now.

File: lib/config/types.ts [hunk 10]
```diff
@@ -357,6 +357,16 @@ export interface RenovateOptionBase {
...
+
+  advancedUse?: boolean;
+
+  followTagAdvancedFunctionality?: boolean;
+
+  groupAdvancedFunctionality?: boolean;
+
+  advancedConfig?: boolean;
+
+  directEditOfCommitMessageDeprecation?: boolean;
```
[the remark this answers is not in the document]
Comment (MaronHatoum) [reply lib/config/types.ts:369]: yes, sure I can return `directEditOfCommitMessageDeprecation` message to the docs because we are using this message just one time.

[the remark this answers is not in the document]
Comment (MaronHatoum) [reply lib/config/types.ts]: maybe we can use the same string in `generateAdvancedUse` and `generateAdvancedConfig`, @HonkingGoose  what do you think?
```
<!-- prettier-ignore -->
!!! warning
    For advanced use only! Use at your own risk!
```

File: lib/config/types.ts [hunk 11]
```diff
@@ -357,6 +357,12 @@ export interface RenovateOptionBase {
   experimentalDescription?: string;
 
   experimentalIssues?: number[];
+
+  advancedUse?: boolean;
+
+  advancedFunctionality?: boolean;
+
+  advancedConfig?: boolean;
```
Comment (rarkins) [inline lib/config/types.ts:365]: can we combine these?

Comment (MaronHatoum) [reply lib/config/types.ts:365]: yes we can if we will use the same string, but if we need 3 different strings I should check the `option name`, and according to this, we can use the relevant string.
something like this:
```
      if (el.advancedUse) {
        if (option.name === 'excludeCommitPaths') {
          configOptionsRaw[headerIndex] += generateAdvancedUse();
        } 
        if(option.name ==='commitMessagePrefix'||option.name ==='commitMessageAction'||
          option.name ==='commitMessageTopic'||option.name ==='commitMessageExtra'||
          option.name ==='commitMessageSuffix'){
          configOptionsRaw[footerIndex] += generateAdvancedUse();
        }
        if(option.name === 'group'|| option.name === 'followTag'){
          configOptionsRaw[headerIndex] += generateAdvancedFunctionality();
        }
        if(option.name === 'insecureRegistry'){
          configOptionsRaw[footerIndex] += generateAdvancedConfig();
        }
      }
```

Comment (rarkins) [reply lib/config/types.ts:365]: I think we can use a simple common admonition text without such complicated logic. If more details are needed for individual options then it can be stated in the markdown documentation separately

Comment (viceice) [reply lib/config/types.ts:365]: That's what i suggested too before 😕

Comment (MaronHatoum) [reply lib/config/types.ts:365]: sounds perfect, what do you recommend for the common text?

Comment (MaronHatoum) [reply lib/config/types.ts:365]:
```
<!-- prettier-ignore -->
!!! warning
For advanced use only! Use at your own risk!
```

can we use this text for all the options?

File: lib/config/options/index.ts [hunk 12]
```diff
@@ -788,7 +788,6 @@ const options: RenovateOptions[] = [
     description:
       'A list of package managers to enable. If defined, then all managers not on the list are disabled.',
     type: 'array',
-    mergeable: false,
```
Comment (viceice) [inline lib/config/options/index.ts:791]: wrong merge?

File: lib/config/types.ts [hunk 13]
```diff
@@ -434,7 +436,7 @@ export interface PackageRuleInputConfig extends Record<string, unknown> {
   depName?: string;
   currentValue?: string | null;
   currentVersion?: string;
-  lockedVersion?: string;
+  lockedVersion?: string | null;
```
Comment (viceice) [inline lib/config/types.ts:439]: wrong merge?

File: tools/docs/config.ts [hunk 14]
```diff
@@ -233,10 +234,29 @@ export async function generateConfig(dist: string, bot = false): Promise<void> {
...
+        if (
+          option.name === 'excludeCommitPaths' ||
+          option.name === 'group' ||
+          option.name === 'followTag'
+        ) {
+          configOptionsRaw[headerIndex] += generateAdvancedUse();
```
Comment (viceice) [inline tools/docs/config.ts:243]: i think we should always add it after the header.

Comment (MaronHatoum) [reply tools/docs/config.ts:243]: in "commitMessages..." the warning text appears at the end right now, we can move it after the table.

Comment (viceice) [reply tools/docs/config.ts:243]: should at least always at same place and very early

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]
